{"componentChunkName":"component---src-templates-blog-post-js","path":"/blog/cpp-error-code/","webpackCompilationHash":"bc0d96a3079a077020f1","result":{"data":{"blog":{"excerpt":"I was recently implementing the “classification of error conditions” in my application offered by the functionality behind std::error_code. In this post I want…","html":"<p>I was recently implementing the “classification of error conditions” in my application offered by the functionality behind <em>std::error_code</em>. In this post I want to share some of my experience and insight.</p>\n<p>C++11 comes with a quite sophisticated mechanism for classifying error conditions. You may have encountered names like “error code”, “error condition”, \"error category”, but figuring out what good they are, and how to use them is difficult. The only valuable source of information on the subject in the Internet is a series of blog posts by Christopher Kohlhoff, the author of Boost.Asio library:</p>\n<ul>\n<li>“<a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-1.html\">System error support in C++0x – part 1</a>”</li>\n<li>“<a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-2.html\">System error support in C++0x – part 2</a>”</li>\n<li>“<a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-3.html\">System error support in C++0x – part 3</a>”</li>\n<li>“<a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-4.html\">System error support in C++0x – part 4</a>”</li>\n<li>“<a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-5.html\">System error support in C++0x – part 5</a>”</li>\n</ul>\n<p>And this was a really good start for me. But still, I believe it would be beneficial to have more than one source of information, and more than one way of explaining the subject. So here we go… </p>\n<p>The problem</p>\n<p>First, why I need it. I have a service for looking for flight connections. You tell me where from and where to you want to go, and I will offer you concrete flights, and a price. In order to be able to do this, my service calls other services in turn:</p>\n<ul>\n<li>one for finding the (short) sequence of flights that will take you to your destination,</li>\n<li>one for checking if there is still seats available on these flights in the requested class of service (economy class, business class)</li>\n</ul>\n<p>Each of these services can fail for a number of reasons. Reasons for failure — different for each service — can be enumerated. For instance the authors of these two services chose the following enumerations:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">enum class FlightsErrc\n{\n    // no 0\n    NonexistentLocations = 10, // requested airport doesn&#39;t exist\n    DatesInThePast,            // booking flight for yesterday\n    InvertedDates,             // returning before departure\n    NoFlightsFound       = 20, // did not find any combination\n    ProtocolViolation    = 30, // e.g., bad XML\n    ConnectionError,           // could not connect to server\n    ResourceError,             // service run short of resources\n    Timeout,                   // did not respond in time\n};\n\nenum class SeatsErrc\n{\n    // no 0\n    InvalidRequest = 1,    // e.g., bad XML\n    CouldNotConnect,       // could not connect to server\n    InternalError,         // service run short of resources\n    NoResponse,            // did not respond in time\n    NonexistentClass,      // requested class does not exist\n    NoSeatAvailable,       // all seats booked\n};</code></pre></div>\n<p>Some things to observe here. First, the reasons for failure are quite similar in either service, but they are assigned different names and different numeric values. This is because two services are developed independently by two different teams. This also means that the same numeric value can refer to two completely different conditions depending on which service reported it.</p>\n<p>Second, as can be seen from the names, the causes for failure come from different sources:</p>\n<ul>\n<li>environment: internal problem in the service (e.g., with resources),</li>\n<li>miscommunication: between two services,</li>\n<li>user: providing incorrect data in request,</li>\n<li>just bad luck: no error actually, but no response can be returned to the user because e.g., all seats have been sold out.</li>\n</ul>\n<p>Now, what do I really need those different error codes for? If any of these errors occurs we want to stop processing the current request from the user. When I can offer no flight connection to him, I only want to distinguish the following situations:</p>\n<ol>\n<li>you gave us illogical request,</li>\n<li>no airline we are aware of is able to offer you a trip,</li>\n<li>there is some problem with the system which you will not understand but which prevents us from giving you the requested answer.</li>\n</ol>\n<p>On the other hand, for the purpose of internal audit, or looking for bugs in the logs, we want a more detailed information to be put into logs, like which system reported the failure, and what actually happened. This can be encoded in an integer number. Any more details, like ports on which we tried to connect, or what database we tried to connect, are likely to be logged separately, so the data encoded in ints should be sufficient.</p>\n<h2>The std::error_code</h2>\n<p>The Standard Library <em>std::error_code</em> is designed to hold exactly this type of information: a number representing the status, and a “domain” within which this number is assigned meaning. In other words, an <em>std::error_code</em> is a pair: <em>{int, domain}</em>. This is reflected in its interface:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">void inspect(std::error_code ec)\n{\n    ec.value();    // the int value\n    ec.category(); // the domain\n}</code></pre></div>\n<p>But you almost never want to inspect an error<em>code this way. As we already said, the two things we want to do is to log the state of the error</em>code as it was constructed (without being later transformed by higher layers of the application), and to use it to answer a specific question, like “is this error caused by the user providing data he knew was incorrect”.</p>\n<p>In case you ask yourself, why use std::error<em>code instead of exceptions, let me clarify: the two things are not mutually exclusive. I want to report failures in my program through exceptions. It is that inside the exception, rather than storing and parsing strings, I just want to contain an error</em>code that I can easily inspect. std::error<em>code has nothing to do with avoiding exceptions. Also, in my use case I do not feel a compelling reason to have many different types of exceptions. I just need one: I will catch them only in one (or two) places and I will tell the different situations by inspecting the error</em>code object.</p>\n<h2>Plugging your enumeration</h2>\n<p>Now, we want to adapt the std::error_code so that it can store error situations from the Flights service described above:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">enum class FlightsErrc\n{\n    // no 0\n    NonexistentLocations = 10, // requested airport doesn&#39;t exist\n    DatesInThePast,            // booking flight for yesterday\n    InvertedDates,             // returning before departure\n    NoFlightsFound       = 20, // did not find any combination\n    ProtocolViolation    = 30, // e.g., bad XML\n    ConnectionError,           // could not connect to server\n    ResourceError,             // service run short of resources\n    Timeout,                   // did not respond in time\n};</code></pre></div>\n<p>We should be able to convert from our enum to std::error_code:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">std::error_code ec = FlightsErrc::NonexistentLocations;</code></pre></div>\n<p>But our enumeration must meet one condition: numeric value 0 must not represent an error situation. 0 represents a success in any error domain (category). This expectation is later exploited when we are inspecting an std::error_code object:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">void inspect(std::error_code ec)\n{\n    if (ec) // equivalent to: ec.value() != 0\n        handle_failure(ec);\n    else\n        handle_success();\n}</code></pre></div>\n<p>In this sense the <a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-4.html\">mentioned blog post</a> has it incorrect that numeric value 200 indicates success.</p>\n<p>So this is what we did: we did not start the enumeration of FlightsErrc with 0. This in turn implies that we can create the enumeration that does not correspond to any of the enumerated values:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">FlightsErrc fe {};</code></pre></div>\n<p>This is an important characteristic of enums in C++ (even the C++11 enum classes): you can create values from outside the enumerated range. It is for this reason that compilers issue a warning in switch-statement that “not all control paths return value” even though you have a case label for every enumeration.</p>\n<p>Now, back to the conversion, std::error_code has a converting constructor template that looks more-less like this:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">template &lt;class Errc&gt;\n    requires is_error_code&lt;Errc&gt;::value\nerror_code(Errc e) noexcept\n    : error_code{make_error_code(e)}\n    {}</code></pre></div>\n<p>(Of course, I used yet non-existent concepts syntax, but you get the idea: this constructor is only visible when std::is<em>error</em>code<Errc>::value evaluates to true.)</p>\n<p>This constructor is meant to be a customization hook for plugging custom error enumerations into the system. In order to plug FlightsErrc, we have to make sure that:</p>\n<ol>\n<li><em>std::is<em>error</em>code<Errc>::value</em> returns true,</li>\n<li>Function <em>make<em>error</em>code</em> taking FlightsErrc is defined and accessible through argument-dependent lookup.</li>\n</ol>\n<p>Regarding the first part, we need to specialize the standard type trait:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">namespace std\n{\n    template &lt;&gt;\n    struct is_error_code_enum&lt;FlightsErrc&gt; : true_type {};\n}</code></pre></div>\n<p>This is one of these situations where declaring something in namespace <em>std</em> is legal.</p>\n<p>Regarding the second part, we just need to declare function overload make<em>error</em>code in the same namespace as enum FlightsErrc:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">enum class FlightsErrc;\nstd::error_code make_error_code(FlightsErrc);</code></pre></div>\n<p>And this is all that other parts of the program/library need to see, and what we have to provide in the header file. The rest is the implementation of function make<em>error</em>code and we can put it in a separate translation unit (a .cpp file).</p>\n<p>With this in place, we can make an impression that FlightsErrc is an error_code:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">std::error_code ec = FlightsErrc::NoFlightsFound;\nassert (ec == FlightsErrc::NoFlightsFound);\nassert (ec != FlightsErrc::InvertedDates);</code></pre></div>\n<h2>Defining error category</h2>\n<p>So far, I have only been saying that error_code is a pair: {number, domain}, where the first element uniquely identifies the particular error situation within the domain, and the second uniquely identifies the domain of errors across all possible error domains that will ever be conceived. But given that this domain ID should be stored in one machine word, how can we guarantee that it will be unique across all libraries currently in the market and those yet to come? We are hiding the domain ID as an implementation detail. If we are to use another third-party library with its own error enumeration, how can we guarantee that their domain ID will not be equal to ours?</p>\n<p>The solution chosen for <em>std::error_code</em> relies on the observation that for every global object (or more formally: namespace-scope object) a unique address is assigned. No matter how many libraries are combined together, with how many globals, each global has a unique address — this is quite obvious.</p>\n<p>In order to exploit this, we have to associate with every type that wants to be plugged into the error<em>code system a unique global object, and then use its address as an ID. Now, this implies using pointers for representing domains, and this is indeed what *std::error</em>code* is doing. But now, that we store some T* the question is what the T should be. The choice is quite clever: let’s use a type that can offer us additional benefits. The type T used is <em>std::error_category</em>, and the additional benefit is in its interface:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">class error_category {\npublic:\n    virtual const char* name() const noexcept = 0;\n    virtual string message(int ev) const = 0;\n    // other members ...\n};</code></pre></div>\n<p>I used name “domain”, the Standard Library uses name “error category” for the same purpose.</p>\n<p>It has pure virtual member functions, which already suggests something: we will be storing pointers to classes derived from std::error<em>category: each new error enum requires a new corresponding class to be derived from std::error</em>category. Usually having pure virtual functions implies allocating objects on the heap, but we will do no such things. We will be creating global objects and pointing to them.</p>\n<p>There are more virtual member functions in std::error_category that on other occasions need to be customized, but we will not have to do it for the purpose of plugging our FlightsErrc.</p>\n<p>Now, for each custom error “domain” represented by a class derived from std::error_category, we need to override two member functions. Function <em>name</em> is expected to return a short mnemonic-like descriptive name of this error category (domain). Function <em>message</em> assigns a text description for every numeric error value in this domain. In order to illustrate it better, let’s define an error category for our enum FlightsErrc. Remember, this class only needs to be visible in one translation unit. In other files we will just be using an address to its instance.</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">namespace { // anonymous namespace\n\nstruct FlightsErrCategory : std::error_category\n{\n    const char* name() const noexcept override;\n    std::string message(int ev) const override;\n};\n\nconst char* FlightsErrCategory::name() const noexcept\n{\n    return &quot;flights&quot;;\n}\n\nstd::string FlightsErrCategory::message(int ev) const\n{\n    switch (static_cast&lt;FlightsErrc&gt;(ev))\n    {\n    case FlightsErrc::NonexistentLocations:\n        return &quot;nonexistent airport name in request&quot;;\n    \n    case FlightsErrc::DatesInThePast:\n        return &quot;request for a date from the past&quot;;\n    \n    case FlightsErrc::InvertedDates:\n        return &quot;requested flight return date before departure date&quot;;\n    \n    case FlightsErrc::NoFlightsFound:\n        return &quot;no filight combination found&quot;;\n    \n    case FlightsErrc::ProtocolViolation:\n        return &quot;received malformed request&quot;;\n    \n    case FlightsErrc::ConnectionError:\n        return &quot;could not connect to server&quot;;\n    \n    case FlightsErrc::ResourceError:\n        return &quot;insufficient resources&quot;;\n    \n    case FlightsErrc::Timeout:\n        return &quot;processing timed out&quot;;\n    \n    default:\n        return &quot;(unrecognized error)&quot;;\n    }\n}\n\nconst FlightsErrCategory theFlightsErrCategory {};\n\n}</code></pre></div>\n<p>Function name provides a short text that is used while streaming out an std::error_code into things like logs: it can help you identify the cause of an error. It does not have to be unique across all different error enums: in the worst case log entries will be ambiguous.</p>\n<p>Function message provides a descriptive message for any numeric value representing an error in our category. This can be helpful in debugging or browsing logs; but you would probably not want to give this text to the users unprocessed. These messages are close to the comments I initially put in the definition of FlightsErrc.</p>\n<p>This function is usually called indirectly. The callers cannot know that the numeric error value is a FlightsErrc, so we have to explicitly cast it back to FlightsErrc. I believe the example in the <a href=\"http://blog.think-async.com/2010/04/system-error-support-in-c0x-part-4.html\">aforementioned article</a> does not compile due to the omitted static<em>cast. Now after the cast, there is a risk that we will be inspecting a value from outside the enumerated set: therefore we need the default label. (Interestingly, whenever I decide to use enum class in my programs, I immediately find myself in need of static</em>casting it either to or from int.)</p>\n<p>Finally, note that we have initialized a global object of our type FlightsErrCategory. This will be the only object of this type in the program. We will need its address (to tell error_codes from different domains), but also we will use its polymorphic properties.</p>\n<p>Although class std::error_category is not a literal type, it has a constexpr default constructor. The implicitly declared default constructor of our class FlightsErrCategory inherits this constexpr-ness. Thus, we are guaranteed that the initialization of our global object is performed during constant initialization, as described in this <a href=\"https://akrzemi1.wordpress.com/2012/05/27/constant-initialization/\">post</a>, and is therefore free from any <a href=\"https://isocpp.org/wiki/faq/ctors#static-init-order\">static initialization order fiasco</a>.</p>\n<p>Now, the last missing part is implementing <em>make<em>error</em>code</em>:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">std::error_code make_error_code(FlightsErrc e)\n{\n    return {static_cast&lt;int&gt;(e), theFlightsErrCategory};\n}</code></pre></div>\n<p>And we are done. Our FlightsErrc can be used as if it was an std::error_code:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">int main()\n{\n    std::error_code ec = FlightsErrc::NoFlightsFound;\n    std::cout &lt;&lt; ec &lt;&lt; std::endl;\n}</code></pre></div>\n<p>The following program will output:</p>\n<div class=\"gatsby-highlight\" data-language=\"text\"><pre class=\"language-text\"><code class=\"language-text\">flights:20</code></pre></div>\n<p>A full working example illustrating all the above can be found <a href=\"https://akrzemi1.wordpress.com/examples/error_code-example/\">here</a>.</p>\n<p>And that’s it for today. What we have still not covered is how to make useful queries on std::error_code objects, but this will be the subject of another post.</p>\n<h2>Acknowledgements</h2>\n<p>I am grateful to Tomasz Kamiński for explaining to me the idea behind std::error<em>code. Apart form Christopher Kohlhoff’s series of posts, I was also able to learn about std::error</em>code form the documentation of Niall Douglas’s <a href=\"https://github.com/ned14/outcome\">Outcome library</a>, <a href=\"https://ned14.github.io/outcome/tutorial/error_code/\">here</a>.</p>\n<p>Andrzej <a href=\"https://akrzemi1.wordpress.com/2017/07/12/your-own-error-code/\">https://akrzemi1.wordpress.com/2017/07/12/your-own-error-code/</a></p>","timeToRead":12,"frontmatter":{"title":"Your own error code","date":"10 Nov 2019","tags":["cpp","error"],"category":"repost","cover":{"childImageSharp":{"fluid":{"base64":"data:image/jpeg;base64,/9j/2wBDABALDA4MChAODQ4SERATGCgaGBYWGDEjJR0oOjM9PDkzODdASFxOQERXRTc4UG1RV19iZ2hnPk1xeXBkeFxlZ2P/2wBDARESEhgVGC8aGi9jQjhCY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2NjY2P/wgARCAALABQDASIAAhEBAxEB/8QAFwAAAwEAAAAAAAAAAAAAAAAAAAIDBf/EABUBAQEAAAAAAAAAAAAAAAAAAAAB/9oADAMBAAIQAxAAAAHGtFlUBP/EABgQAQADAQAAAAAAAAAAAAAAAAEAAhAD/9oACAEBAAEFAp0qV1Vz/8QAFREBAQAAAAAAAAAAAAAAAAAAEBH/2gAIAQMBAT8Bp//EABQRAQAAAAAAAAAAAAAAAAAAABD/2gAIAQIBAT8BP//EABkQAAMAAwAAAAAAAAAAAAAAAAEQEQAhgf/aAAgBAQAGPwLBL17K/8QAGxAAAgEFAAAAAAAAAAAAAAAAARAAESExUXH/2gAIAQEAAT8hmywvQ8gHpX//2gAMAwEAAgADAAAAEOPP/8QAFhEBAQEAAAAAAAAAAAAAAAAAAREQ/9oACAEDAQE/ECBJn//EABQRAQAAAAAAAAAAAAAAAAAAABD/2gAIAQIBAT8QP//EABoQAAICAwAAAAAAAAAAAAAAAAERECEAMWH/2gAIAQEAAT8Qwgw7WBfJQUemFR//2Q==","aspectRatio":1.7777777777777777,"src":"/static/880e45de7382d435ec205486253e821d/67481/error.jpg","srcSet":"/static/880e45de7382d435ec205486253e821d/b6231/error.jpg 200w,\n/static/880e45de7382d435ec205486253e821d/f4032/error.jpg 400w,\n/static/880e45de7382d435ec205486253e821d/67481/error.jpg 800w,\n/static/880e45de7382d435ec205486253e821d/c4f3a/error.jpg 1200w,\n/static/880e45de7382d435ec205486253e821d/f232b/error.jpg 1600w,\n/static/880e45de7382d435ec205486253e821d/654ed/error.jpg 1920w","sizes":"(max-width: 800px) 100vw, 800px"}}}},"fields":{"slug":"/blog/cpp-error-code/"}}},"pageContext":{"isCreatedByStatefulCreatePages":false,"slug":"/blog/cpp-error-code/","previous":{"fields":{"slug":"/blog/leveldb/"}},"next":{"fields":{"slug":"/blog/nosql-databases-an-overview/"}}}}}